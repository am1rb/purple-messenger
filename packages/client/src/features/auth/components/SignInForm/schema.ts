import * as yup from "yup";

function demoValueOrSchema(schema: yup.StringSchema) {
  return yup.lazy((value: string) =>
    value !== "demo"
      ? yup
          .string()
          .email()
          .required()
      : yup.string()
  );
}

export default yup.object().shape({
  email: demoValueOrSchema(
    yup
      .string()
      .email()
      .required()
  ),
  password: demoValueOrSchema(
    yup
      .string()
      .min(8)
      .required()
  )
});
