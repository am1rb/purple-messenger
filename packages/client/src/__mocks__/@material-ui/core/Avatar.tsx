import React from "react";
import type { AvatarProps } from "@material-ui/core/Avatar";

function MuiAvatar({ src, alt }: AvatarProps) {
  return <img data-testid="mock-mui-avatar" src={src} alt={alt} />;
}

export default MuiAvatar;
