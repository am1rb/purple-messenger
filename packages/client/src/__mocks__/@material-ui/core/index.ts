export { default as Avatar } from "./Avatar";
export { default as Box } from "./Box";
export { default as Card } from "./Card";
export { default as CardContent } from "./CardContent";
export { default as Typography } from "./Typography";
export { default as Button } from "./Button";
