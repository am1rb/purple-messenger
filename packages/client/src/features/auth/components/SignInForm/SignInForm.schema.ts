import * as yup from "yup";

function demoValueOrSchema(schema: yup.StringSchema<string>) {
  return yup.lazy((value: string) => 
    value !== "demo"
      ? schema
      : yup.string().required()
  );
}

export default yup.object().shape({
  email: demoValueOrSchema(
    yup
      .string()
      .email()
      .required()
  ),
  password: demoValueOrSchema(
    yup
      .string()
      .min(8)
      .required()
  )
});
