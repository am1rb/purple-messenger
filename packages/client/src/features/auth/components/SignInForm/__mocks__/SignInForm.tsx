import React from "react";

function SignInForm() {
  return <form data-testid="mock-sign-in-form" />;
}

export default SignInForm;
