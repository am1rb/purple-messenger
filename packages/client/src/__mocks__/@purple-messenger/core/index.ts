import * as purpleMessengerCore from "@purple-messenger/core";

const mockedPurpleMessengerCore = jest.genMockFromModule<
  typeof purpleMessengerCore
>("@purple-messenger/core");

function mockedAction(action = () => ({ type: "" })) {
  return jest.fn(action);
}

module.exports = {
  ...mockedPurpleMessengerCore,
  loadConversationList: mockedAction(),
  unloadConversationList: mockedAction(),
  signOut: mockedAction(),
  clearMessageList: mockedAction(),
  seenMessageAck: mockedAction(),
  start: mockedAction(),
  shutdown: mockedAction(),
  submitMessage: mockedAction(),
  startTypingMessage: mockedAction(),
  stopTypingMessage: mockedAction(),
  signIn: mockedAction(),
  setAuthError: mockedAction(),
  verifyToken: mockedAction(),
  connected: mockedAction(purpleMessengerCore.connected),
  disconnected: mockedAction(purpleMessengerCore.disconnected),
};
