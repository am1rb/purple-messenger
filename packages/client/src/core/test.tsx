import React from "react";
import { render } from "@testing-library/react";
import { Provider } from "react-redux";
import { ConnectedRouter } from "connected-react-router";
import configureStore, { history } from "core/redux/configureStore";
import { State as RootState } from "core/redux/reducers";
import theme from "core/style/theme";
import { ThemeProvider } from "@material-ui/styles";

export function renderWithStore(
  ui: React.ReactElement,
  initStore: Partial<RootState>
) {
  const store = configureStore(initStore);

  return render(ui, {
    wrapper: ({ children }) => (
      <ThemeProvider theme={theme}>
        <Provider store={store}>
          <ConnectedRouter history={history}>{children}</ConnectedRouter>
        </Provider>
      </ThemeProvider>
    ),
  });
}
