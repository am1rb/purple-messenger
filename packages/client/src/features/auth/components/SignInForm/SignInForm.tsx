import React, { memo, useCallback } from "react";
import { useDispatch, useSelector } from "react-redux";
import yup from "yup";
import { Button } from "@material-ui/core";
import Unform from "components/Unform";
import UnformTextField from "components/UnformTextField";
import { SignInFormContent } from "../../types/formContent";
import { signIn, setAuthError } from "@purple-messenger/core";
import { getAuthError } from "features/auth/selectors";
import schema from "./SignInForm.schema";
import useStyles from "./SignInForm.styles";

function SignInForm() {
  const classes = useStyles();
  const dispatch = useDispatch();
  const authError = useSelector(getAuthError);

  const handleSubmit = useCallback(
    (content: SignInFormContent) => {
      dispatch(setAuthError(""));
      dispatch(signIn(content.email, content.password));
    },
    [dispatch]
  );

  return (
    <Unform<SignInFormContent>
      schema={schema as yup.ObjectSchema<object>}
      onSubmit={handleSubmit}
      className={classes.form}
    >
      {authError}
      <UnformTextField name="email" label="Email" margin="normal" />
      <UnformTextField
        name="password"
        type="password"
        label="Password"
        autoComplete="current-password"
        margin="normal"
      />
      <Button type="submit" variant="contained" color="primary">
        SignIn
      </Button>
    </Unform>
  );
}

export default memo(SignInForm);
