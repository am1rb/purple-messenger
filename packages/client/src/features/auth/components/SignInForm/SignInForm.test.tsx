import React from "react";
import userEvent from "@testing-library/user-event";
import { signIn, setAuthError } from "@purple-messenger/core";
import { renderWithStore } from "core/test";
import SignInForm from "./SignInForm";

jest.mock("components/Unform");
jest.mock("components/UnformTextField");

describe("The <SignInForm /> tests", () => {
  it("Should match the snapshot", () => {
    const { container } = renderWithStore(<SignInForm />);
    expect(container).toMatchSnapshot();
  });

  it("Should call signIn action when the form is submitted", () => {
    const username = "john@doe.com";
    const password = "strong-password";

    const { getByLabelText, getByTestId, store } = renderWithStore(
      <SignInForm />
    );

    userEvent.type(getByLabelText("Email"), username);
    userEvent.type(getByLabelText("Password"), password);

    userEvent.click(getByTestId("mock-mui-button"));

    expect(store.actions).toContainEqual(setAuthError(""));
    expect(store.actions).toContainEqual(signIn(username, password));
  });
});
