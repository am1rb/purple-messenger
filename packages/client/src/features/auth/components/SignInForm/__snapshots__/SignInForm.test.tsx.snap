// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The <SignInForm /> tests Should match the snapshot 1`] = `
<div>
  <form>
    
    <label>
      Email
      <input
        data-testid="mock-unform-text-field"
        margin="normal"
        name="email"
      />
    </label>
    <label>
      Password
      <input
        autocomplete="current-password"
        data-testid="mock-unform-text-field"
        margin="normal"
        name="password"
        type="password"
      />
    </label>
    <button
      data-testid="mock-mui-button"
    >
      SignIn
    </button>
  </form>
</div>
`;
